Bootstrap package requests

bgeuken opened the pull request.

hennevogel commented on src/api/app/views/webui/shared/_requests_state_and_type_form.html.haml:
```diff
@@ -0,0 +1,12 @@
+               data: { table: 'all_requests_table' })
+  %span#request_count
+  \:
+  = image_tag('ajax-loader.gif', id: 'spinner', class: 'hidden')
```
This is [`.invisible`](https://getbootstrap.com/docs/4.1/utilities/visibility/) in bootstrap...

hennevogel commented:
I can give it a shot if you want to...

bgeuken commented:
* Rebased with the other request related changes (js and haml). That previously broke the spinner.
* Added font awesome spinner icons
* Wrapped the tab content in a card. That also fixed the responsiveness issue.

hennevogel commented on src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml:
```diff
@@ -0,0 +1,13 @@
+%p
+  Display
+  = select_tag('request_type_select',
```
You forgot the `.form-control` class. Also the whole thing is rather crowded and unaligned. 

![screen shot 2018-08-22 at 00 45 59](https://user-images.githubusercontent.com/514785/44433015-d31e5480-a5a4-11e8-8d62-6435fcf6d8d8.png)

hennevogel commented on src/api/app/views/webui2/shared/_requests_table.html.haml:
```diff
@@ -0,0 +1,13 @@
+%table.requests-datatable.table.table-striped.table-bordered{ 'data-source': source_url,
```
There are [responsive tables](https://getbootstrap.com/docs/4.1/content/tables/#responsive-tables) in bootstrap, there is also an [datatables extension](https://datatables.net/extensions/responsive/) to do this, with bootstrap integration. Check it out, or get a card in the backlog about this going please.

mdeniz commented:
Something strange is happening to the table in full size:

![image](https://user-images.githubusercontent.com/11314634/44461284-807a8200-a610-11e8-86f9-81e47a356f01.png)

bgeuken commented:
I am still working on updating the view

mdeniz commented:
Hey, whenever you are done with this PR... close it and open another to the `bootstrap-master` branch instead.

bgeuken commented:
Why bootstrap-master?

mdeniz commented:
> Why bootstrap-master?

Remember that is in that branch where we are going to add changes for bootstrap instead of master. The point was to have smaller PRs that may break the workflows and until a page/workflow is fully working it shouldn't be merged on master as we deploy that on our instance.

bgeuken commented:
I consider this PR quite small and it's isolated from the old webui workflow and ui

mdeniz commented:
> I consider this PR quite small and it's isolated from the old webui workflow and ui

Ok, keep it that way. But, for the next ones we must to follow our own agreements :+1:

bgeuken commented:
> mdeniz commented 
> I consider this PR quite small and it's isolated from the old webui workflow and uiOk, keep it that way. But, for the next ones we must to follow our own agreements

Well, I talked with Henne before and he confirmed. So I follow our agreements

Or to cite the initial mail regarding this:
`We can only merge complete working pages into master`

And this is the case here:+1:

bgeuken commented:
@hennevogel @mdeniz @dmarcoux Updated the PR again. Took a bit longer than I hoped. Sorry for that.

One remaining issue:
* The 'Show x entries' and Search bar are getting centered at a certain (small enough) screen size.
  That's because both elements are generated by datatable, while the two select boxes are customized.

I find that a bit annoying, but I also don't want to write lot's custom css for this. Since it's kind of an edge case

dmarcoux commented on src/api/app/views/webui2/webui/package/requests.html.haml:
```diff
@@ -0,0 +1,14 @@
+- @pagetitle = "Requests for #{@package}"
+- package_bread_crumb('Requests')
```
The line should be removed now that it is in the breadcrumbs partial.

bgeuken replied:
Oh, thanks. I am blind:blush:

hennevogel commented:
> I also don't want to write lot's custom css for this

Well we can create a card for this, no problem.

hennevogel commented on src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml:
```diff
@@ -0,0 +1,16 @@
+      = select_tag('request_state_select',
+                   options_for_select(Webui::RequestHelper::AVAILABLE_STATES, selected_state),
+                   data: { table: 'all_requests_table' },
+                   class: 'form-control custom-select custom-select-sm form-control-sm dataTables_length')
```
Now this looks a bit strange on the larger screen. 

![screenshot from 2018-08-22 17-13-20](https://user-images.githubusercontent.com/514785/44472742-11605600-a62f-11e8-8df1-4ed38e87d767.png)

How about you move the selects into partials and then you can render them more than one time and [hide](https://getbootstrap.com/docs/4.1/utilities/display/#hiding-elements) what you don't need in the different breakpoints.

bgeuken replied:
Updated. I found the CSS for setting correct breakpoints for different screensizes. So no need for hiding and showing elements:tada:

bgeuken commented:
If you know of a better icon for the request links let me know:wink:

hennevogel commented on src/api/app/controllers/webui/packages/bs_requests_controller.rb:
```diff
@@ -13,7 +13,11 @@ def index
+            format.json { render 'webui2/shared/bs_requests/index' }
+          else
+            format.json
+          end
```
Interesting way of doing this :)

bgeuken replied:
The `switch_to_webui`  doesn't seem to work when using a respond_to block:/

hennevogel commented on src/api/spec/support/capybara.rb:
```diff
@@ -12,6 +12,7 @@
     browser_options = ::Selenium::WebDriver::Chrome::Options.new
     browser_options.args << '--headless'
     browser_options.args << '--no-sandbox' # to run in docker
+    browser_options.args << '--window-size=1280,1024'
```
What was it before?

bgeuken replied:
I couldn't find that in the documentation, but it looked like ~500px

hennevogel commented on src/api/app/views/webui2/shared/bs_requests/index.json.erb:
```diff
@@ -0,0 +1,23 @@
+          "<%= escape_javascript(user_with_realname_and_icon(row.creator, short: true)) %>",
+          "<%= escape_javascript(new_or_update_request(row)) %>",
+          "<%= escape_javascript(row.priority) %>",
+          "<%= escape_javascript(link_to(icon('fa', 'info-circle'), request_show_path(row.number), { title: "Show request ##{row.number}", class: :request_link })) %>"
```
Oh please avoid the bloody helpers if you can. I would really appreciate it if we could get rid of all this nonsense and pull only over what we really need...

But we can do this later, I'm fine with a FIXME...

bgeuken replied:
The `icon` call? What's so bad about it?

bgeuken replied:
I see, though [icon](https://github.com/FortAwesome/font-awesome-sass#rails-helper-usage) is a at least not a custom one.

dmarcoux commented:
Looking at a page with no requests, I find a few elements a bit strange.

- The text `Showing 0 to 0 of 0 entries`. That's a lot of zeros... I would hide this when there are no entries.
- The pagination links `First`, `Next` (etc...) also look out of place in this context. I'm not sure if hiding them is the way to go here. Just wanted to state that I find this weird too.
- The last column in the datatable after `Priority` does not have a name and it looks like something is wrong with the datatable when there is no data. Maybe add some text like `Actions`

![no_data](https://user-images.githubusercontent.com/1102934/44574389-965f8280-a789-11e8-873e-ed3b3092e8ea.png)

bgeuken commented:
I think that's fine. It's the default datatables behaviour and the pagination buttons are disabled when there is no previous / next request page.
Also I would find it a bit distracting when these elements appear and disappear all the time.

To me a  'Showing 0 of 0 entries' would still be useful since it tells me that there are no results for my request state / type selection.

bgeuken commented:
Regarding the priority sorting. That works (just tested on my machine). In the review app it does not because both requests have the same priority.

And for the action label... I would be fine with that. But I would prefer to do that in a follow up PR. Right now this is like it was in the old UI and changing that might mess up the column size.
